packages/gatsby/src/utils/websocket-manager.ts

```diff
@@ -176,22 +151,39 @@ export class WebsocketManager {
       }
 
       const getDataForPath = async (path: string): Promise<void> => {
-        if (!this.pageResults.has(path)) {
+        let pageData = this.pageResults.get(path)
+        if (!pageData) {
           try {
-            const result = await getCachedPageData(path)
+            pageData = await getPageData(path)
 
-            this.pageResults.set(path, result)
+            this.pageResults.set(path, pageData)
           } catch (err) {
             console.log(err.message)
-
             return
           }
         }
```

Comment:
Does this have the potential for a race condition? (If so it was already there, but still)

If `getDataForPath` gets called twice for the same page in parallel then the second call will not find the result in the cache and not be aware that a "get" for this page is already in flight, effectively fetching the same page twice.

If that is indeed a concern (I'm not certain it is, it does look this way to me) then I'd suggest something like this.

```suggestion
let promise = this.pageResults.get(path)
let pageData
if (promise) {
  try {
    pageData = await promise
  } catch {
    // Ignore. Do not emit duplicate errors for same get.
  }
} else {
  try {
    const promise = getPageData(path)
    this.pageResults.set(path, promise) // Set before attempting to resolve
    pageData = await promise
  } catch (err) {
    // Error stuff. Once per path.
  }
}
```

The `pageResults` don't show up in other places in this diff but if they are used then those places should also update to take the promise into account. Also not sure whether this is sufficient to deal with the double error. Maybe it could be consolidated into the same code path by tracking whether an error was thrown for this promise, rather than the above.

Considering this problem already exists, perhaps it's good to do in a followup. But I think it ought to be checked for sure.